### PR TITLE
Refactor MultiMatmulSchedulers

### LINKED ISSUE #3266: Refactor MultiMatrixScheduler
1. Move `isPowOf2` to `csrc/utils.h`
2. Move `representativeId` to `scheduler/tools/abstract_tensor.h`
3. Move `checkConcreteStaticDim` to `mma_utils.cpp`
4. Remove `swizzleSharedMemory` from `HopperMultiMatrixScheduler`
5. Create base class `MultiMatrixScheduler` to hold common functions like `findPatterns`

### PR DESCRIPTION
**Pull request by rdspring1**
The PR fixes #3266. 

* Future consolidation is possible but I kept some duplicate functions in `HopperMultiMatrixScheduler` and `AmpereMultiMatrixScheduler` for flexibility.

### Changes:
- Move `isPowOf2` to `csrc/utils.h`
- Move `representativeId` to `scheduler/tools/abstract_tensor.h`
- Move `checkConcreteStaticDim to mma_utils.cpp`
- Add TODO to remove `swizzleSharedMemory` from `HopperMultiMatrixScheduler`
- Create base class `MultiMatrixScheduler` to hold common functions like `findPatterns`

### Details:
- Create base class `MultiMatrixScheduler`
- `HopperMultiMatrixScheduler` and `AmpereMultiMatrixScheduler` inherit from `MultiMatrixScheduler` and overwrite the `run` function.
-  `MultiMatrixScheduler` implements `findPatterns`, `translatePatterns`, `findRoles`, `countDims`, and `updateIdModel`. It also holds the necessary data members for those functions.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

File: csrc/scheduler/multi_matmul.h [hunk 1]
```diff
@@ -9,10 +9,64 @@
...
+#include <scheduler/mma_utils.h>
+#include <val_graph.h>
+#include <val_graph_visitor.h>
 #include <visibility.h>
 
 namespace nvfuser {
 
+// Base class for AmpereMultipleMatmulScheduler and
+// HopperMultipleMatmulScheduler
+class MultipleMatmulScheduler {
```
Comment (jacobhinkle) [inline csrc/scheduler/multi_matmul.h:21]: It could be in another PR but we could probably remove the `Multiple` from the names of these classes. It was needed while we had two matmul schedulers temporarily but not any more.

File: csrc/scheduler/multi_matmul.h [hunk 2]
```diff
@@ -9,10 +9,64 @@
...
+  mma_utils::DimRolesMap id_roles_;
+  mma_utils::TensorRolesMap tensor_roles_;
+  mma_utils::MatmulOperandInnerDims inner_dims_;
+
+  int64_t num_splitk_dims_ = 0;
+  int64_t num_device_dims_ = 0;
+  int64_t num_local_batch_dims_ = 0;
+  int64_t num_device_and_batch_dims_ = 0;
+
+  std::vector<TensorView*> as_, bs_, mma_results_;
```
Comment (jacobhinkle) [inline csrc/scheduler/multi_matmul.h:67]: Good thinking to not include all of these in the parent class.

**Review by jacobhinkle — APPROVED** [review]
LGTM

File: csrc/scheduler/ampere_multi_matmul.h [hunk 3]
```diff
@@ -66,37 +64,23 @@ namespace nvfuser {
 // Each of the named tensors above is scheduled differently. We schedule them
 // by building AbstractTensors for each tensor category; these are held in
 // AmpereMultipleMatmulScheduler::schedules_.
-// TODO: Inheret from SchedulerEntry
-class AmpereMultipleMatmulScheduler {
+// TODO: Inherit from SchedulerEntry
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/scheduler/ampere_multi_matmul.h:67]: Perhaps. The matmul parameters per generation can be different.

**Comment by rdspring1** [comment]
!build
